### PR TITLE
Establishment of a Requested Stock % feature, and automating weekly checks to keep spare parts in stock

### PR DESCRIPTION
**Pull request by Lapras**
Making sure your mercenary company has all the spares needed to contribute to ongoing repairs is a massive manual endeavor once you have anything more than a Lance. This PR adds features to the Parts In Use report for the user. Allow them to request a percentage of stock of a part in use that they want to keep in stock. For example, if a company is using 20 medium lasers, and your stock percentage is 25%, you are asking to always have 5 spare medium lasers on hand for repairs. It adds a button that will compare current stock to requested stock and automatically put in orders (or add them via GM mode) to get you all those parts. Finally, there's a checkbox to make this check every Monday. Finally, your logistics officers can do their jobs instead of you having to direct them to buy every part necessary.

This is done with a couple of changes:

1. In CampaignOptions, a default stock level percentage is added to the options that is applied to every new part in use by default
2. We need to store our settings for parts in use now (ignore mothballed, minimum part quality, weekly check, and all of the stock percentages), so we've added some marshalling and unmarshalling in the XML so that saved games now track what parts in use are requested at what levels. Keep in mind that this is done by storing a simple map from PIU names to doubles, we don't actually store all of the information contained within the parts in use report, just a map from names to the levels, and reapply them when parts in use are generated.
3. In the Campaign.java, we're keeping track of our partInUse options and have implemented functions to both purchase and add parts that aren't in stock.
4. The Parts In Use dialog has a column added to it for the request stock % as well as buttons to fill stock that access these campaign functions
5. A weekly check on newDay() has been added in campaign in order to check your stock levels weekly
6. Parts In Use no longer derive their name (and therefore how they're compared) by the description of their part, but rather the AcquisitionName of the IAcquisitionWork object that is their partToBuy, this gives us a more consistent naming scheme and avoids problems with part in use recognition. I've added in a special edge case for turrets, as their tonnage isn't specified within their IAcquisitonWork but does matter.

### CODE REVIEW

**Comment by IllianiCBT** [comment]
@HammerGS i want the opportunity to review this before it gets merged, but it needs to be merged before CO IIC else it’ll create a massive pain of conflicts.

**Comment by HammerGS** [comment]
> @HammerGS i want the opportunity to review this before it gets merged, but it needs to be merged before CO IIC else it’ll create a massive pain of conflicts.

I'll let you handle the review and merge

**Comment by IllianiCBT** [comment]
Looks like this has picked up a conflict. @Sleet01 I don't suppose you'll have time to cast your eyes over this as well? With this being a high profile addition that's going to get a lot of attention I wanna make sure we're catching any issues before launch.

File: MekHQ/resources/mekhq/resources/CampaignOptionsDialog.properties [hunk 1]
```diff
@@ -76,7 +76,9 @@ lblAcquireSkill.text=Acquisition Skill:
 lblSupportStaffOnly.text=Only support personnel can make acquisition checks
 lblWaitingPeriod.text=Waiting period (in days) between acquisition rolls
 spnMaxAcquisitions.toolTipText=This number is the maximum number of acquisitions an administrator can attempt each day. Setting this to 0 makes it unlimited.
+spnDefaultStockPercent.toolTipText=This number is the default % of stock to fill for parts in use
```
Comment (IllianiCBT) [inline MekHQ/resources/mekhq/resources/CampaignOptionsDialog.properties:79]: Minor stylistic choice, can we change `%` to `percent` or `percentage`

File: MekHQ/resources/mekhq/resources/PartsInUseTableModel.properties [hunk 2]
```diff
@@ -10,3 +10,4 @@ ordered.heading=Ordered
 part.heading=Part
 stored.heading=Stored
 storedTonnage.heading=Tonnage
+requestedStock.heading =Requested Stock Percent
```
Comment (IllianiCBT) [inline MekHQ/resources/mekhq/resources/PartsInUseTableModel.properties:13]: Extra whitespace after `.heading`

File: MekHQ/src/mekhq/campaign/Campaign.java [hunk 3]
```diff
@@ -174,6 +174,8 @@ public class Campaign implements ITechManager {
     private final TreeMap<Integer, Scenario> scenarios = new TreeMap<>();
     private final Map<UUID, List<Kill>> kills = new HashMap<>();
 
+    private Map<String, Double> piuStockMap = new LinkedHashMap<>(); // This map keeps track of which parts i use are assigned which stock values
```
Comment (IllianiCBT) [inline MekHQ/src/mekhq/campaign/Campaign.java:177]: This comment is a little hard to parse. Is there a typo?

File: MekHQ/src/mekhq/campaign/Campaign.java [hunk 4]
```diff
@@ -271,6 +273,10 @@ public class Campaign implements ITechManager {
     private FameAndInfamyController fameAndInfamy;
     private List<Unit> automatedMothballUnits;
 
+     //options relating to parts in use and restock
+    private boolean ignoreMothballed, topUpWeekly;
```
Comment (IllianiCBT) [inline MekHQ/src/mekhq/campaign/Campaign.java:277]: Having a twinned variable isn't inherently wrong, but it's not something I've seen elsewhere in our codebase. Strictly personnel preference: I'd define these separately. However, I want to emphasize that this is personal preference so go with whatever your heart tells you. :)

File: MekHQ/src/mekhq/campaign/Campaign.java [hunk 5]
```diff
@@ -2236,6 +2245,7 @@ private PartInUse getPartInUse(Part p) {
             p = ((MissingPart) p).getNewPart();
```
Comment (IllianiCBT) [inline MekHQ/src/mekhq/campaign/Campaign.java:2245]: I know this isn't your code, but we're trying to move away from single letter or abbreviated variable names. As you're doing a lot of work in this neck of the woods would you mind changing `p` to `part`?

File: MekHQ/src/mekhq/campaign/Campaign.java [hunk 6]
```diff
@@ -2314,6 +2324,11 @@ public Set<PartInUse> getPartsInUse(boolean ignoreMothballedUnits,
             if (inUse.containsKey(partInUse)) {
                 partInUse = inUse.get(partInUse);
             } else {
+                if(piuStockMap.containsKey(partInUse.getDescription())) {
```
Comment (IllianiCBT) [inline MekHQ/src/mekhq/campaign/Campaign.java:2327]: Can we come up with a better variable name than `piuStockMap` maybe `partsInUseStockMap`, but I'm sure you can come up with something better. I get that this is annoying, but I've lost a lot of time trying to decipher single letter or abbreviated variable names that were probably super obvious when the author wrote them, but a decade later have lost all meaning.

File: MekHQ/src/mekhq/campaign/Campaign.java [hunk 7]
```diff
@@ -4327,6 +4347,11 @@ public boolean newDay() {
             }
         }
 
+        if(topUpWeekly == true && getLocalDate().getDayOfWeek() == DayOfWeek.MONDAY) {
+            int bought = stockUpPartsInUse(getPartsInUse(ignoreMothballed, ignoreSparesUnderQuality));
+            addReport("Weekly Stock Check: " + bought + " different kinds of parts are under requested stock levels, orders have been put in to buy them");
```
Comment (IllianiCBT) [inline MekHQ/src/mekhq/campaign/Campaign.java:4352]: Can we get this shifted over to the campaign resource bundle

File: MekHQ/src/mekhq/campaign/Campaign.java [hunk 8]
```diff
@@ -8423,4 +8454,96 @@ public String getCommanderAddress() {
         return commanderRank;
     }
 
+    public int stockUpPartsInUse(Set<PartInUse> partsInUse)  {
+        int bought = 0;
+        for(PartInUse piu : partsInUse) {
```
Comment (IllianiCBT) [inline MekHQ/src/mekhq/campaign/Campaign.java:8459]: Can we also change `piu` to non-abbreviated variable name

File: MekHQ/src/mekhq/campaign/Campaign.java [hunk 9]
```diff
@@ -8423,4 +8454,96 @@ public String getCommanderAddress() {
...
+                IAcquisitionWork partToBuy = piu.getPartToBuy();
+                getShoppingList().addShoppingItem(partToBuy, toBuy, this);
+            }
+            bought += 1;
+        }
+        return bought;
+    }
+
+    public void stockUpPartsInUseGM(Set<PartInUse> partsInUse)  {
+        for(PartInUse piu : partsInUse) {
```
Comment (IllianiCBT) [inline MekHQ/src/mekhq/campaign/Campaign.java:8471]: Abbreviated variable name

File: MekHQ/src/mekhq/campaign/Campaign.java [hunk 10]
```diff
@@ -8423,4 +8454,96 @@ public String getCommanderAddress() {
...
+                IAcquisitionWork partToBuy = piu.getPartToBuy();
+                getQuartermaster().addPart((Part) partToBuy.getNewEquipment(), 0);
+                -- toBuy;
+            }
+        }
+    }
+
+    private int findStockUpAmount(PartInUse piu) {
+        IAcquisitionWork partToBuy = piu.getPartToBuy();
+        int inventory = piu.getStoreCount() + piu.getTransferCount() + piu.getPlannedCount();
```
Comment (IllianiCBT) [inline MekHQ/src/mekhq/campaign/Campaign.java:8483]: abbreviated variable name

File: MekHQ/resources/mekhq/resources/PartsInUseTableModel.properties (hunk 2 above)
Comment (IllianiCBT) [reply MekHQ/resources/mekhq/resources/PartsInUseTableModel.properties:13]: In fact, I've noticed a lot of weird whitespacing throughout your code. Lots of `if(` or `while(` instead of `if (` and `while (` which is more standard for our codebase. I'm not going to force you to change them, but I do think we need to maintain whitespace consistency where possible.

File: MekHQ/src/mekhq/campaign/Campaign.java [hunk 11]
```diff
@@ -8423,4 +8454,96 @@ public String getCommanderAddress() {
...
+        int needed = (int)Math.ceil(piu.getRequestedStock()/100.0 * piu.getUseCount());
+        int toBuy = needed-inventory;
+
+        if(piu.getIsBundle() == true) {
+            toBuy = (int)Math.ceil((float)toBuy * piu.getTonnagePerItem() / 5);
+            //special case for ammo, need to track down if there's a way to code this properly
```
Comment (IllianiCBT) [inline MekHQ/src/mekhq/campaign/Campaign.java:8489]: In my experience, there isn't. However, can I verify that you're wanting to buy ammo in 0.2t blocks?

File: MekHQ/src/mekhq/campaign/Campaign.java [hunk 12]
```diff
@@ -8423,4 +8454,96 @@ public String getCommanderAddress() {
...
+            System.out.println("TPI: " + piu.getTonnagePerItem() + " " + String.format("Inv: %d needed: %d tobuy: %d", inventory, needed, toBuy));
+            System.out.println("||");
+        }
+
+       
+        return toBuy;
+    }
+
+    //Simple getters and setters for our stock map
+    public Map<String,Double> getPiuStockMap() {
```
Comment (IllianiCBT) [inline MekHQ/src/mekhq/campaign/Campaign.java:8502]: You're probably fed up of seeing me write this by now, but we don't want abbreviated method names either. Especially for public methods.

File: MekHQ/src/mekhq/campaign/Campaign.java [hunk 13]
```diff
@@ -8423,4 +8454,96 @@ public String getCommanderAddress() {
...
+
+    public PartQuality getIgnoreSparesUnderQuality() {
+        return ignoreSparesUnderQuality;
+    }
+    public void setIgnoreSparesUnderQuality(PartQuality ignoreSparesUnderQuality) {
+        this.ignoreSparesUnderQuality = ignoreSparesUnderQuality;
+    }
+    
+
+    public void writePiuToXML(final PrintWriter pw, int indent) {
```
Comment (IllianiCBT) [inline MekHQ/src/mekhq/campaign/Campaign.java:8532]: Let's get away from abbreviated method names, please :)

File: MekHQ/src/mekhq/campaign/io/CampaignXmlParser.java [hunk 14]
```diff
@@ -772,7 +775,7 @@ private static void processLanceNodes(Campaign retVal, Node wn) {
             }
 
             if (!wn2.getNodeName().equalsIgnoreCase("lance")) {
-                // Error condition of sorts!
+                // Error condition of [sorts]!
```
Comment (IllianiCBT) [inline MekHQ/src/mekhq/campaign/io/CampaignXmlParser.java:778]: What is this change in relation to?

File: MekHQ/src/mekhq/campaign/io/CampaignXmlParser.java [hunk 15]
```diff
@@ -1593,6 +1596,81 @@ private static void updatePlanetaryEventsFromXML(Node wn) {
...
+            }
+
+            processPiuStockMapVal(retVal, wn2, piuStockMap);
+
+        }
+
+        retVal.setPiuStockMap(piuStockMap);
+    }
+
+    private static void processPiuStockMapVal(Campaign retVal, Node wn, Map<String, Double> piuStockMap) {
```
Comment (IllianiCBT) [inline MekHQ/src/mekhq/campaign/io/CampaignXmlParser.java:1650]: Abbreviations

File: MekHQ/src/mekhq/gui/dialog/PartsReportDialog.java [hunk 16]
```diff
@@ -344,6 +402,65 @@ private void refreshOverviewPartsInUse() {
...
+    }
+
+    private int findTopUpAmount(PartInUse piu) {
+        IAcquisitionWork partToBuy = piu.getPartToBuy();
+        int inventory = piu.getStoreCount() + piu.getTransferCount() + piu.getPlannedCount();
+        int needed = (int)Math.ceil(piu.getRequestedStock()/100.0 * piu.getUseCount());
+        int toBuy = needed-inventory;
+
+        if(piu.getIsBundle() == true) {
+            toBuy = (int)Math.ceil((float)toBuy * piu.getTonnagePerItem() / 5);
```
Comment (IllianiCBT) [inline MekHQ/src/mekhq/gui/dialog/PartsReportDialog.java:433]: Abbreviations and same question as before: are we meaning to buy 0.2t of ammo?

File: MekHQ/src/mekhq/gui/dialog/PartsReportDialog.java [hunk 17]
```diff
@@ -344,6 +402,65 @@ private void refreshOverviewPartsInUse() {
...
+            //special case for ammo, need to track down if there's a way to code this properly
+        }
+
+        if(toBuy > 0) {
+            System.out.println("TPI: " + piu.getTonnagePerItem() + " " + String.format("Inv: %d needed: %d tobuy: %d", inventory, needed, toBuy));
```
Comment (IllianiCBT) [inline MekHQ/src/mekhq/gui/dialog/PartsReportDialog.java:438]: Let's definitely push this to logger rather than System.out. System.out is fine for temporary logging, _maybe_, but for anything like this we need a record of it when players find issues and send in their logs.

File: MekHQ/src/mekhq/gui/model/PartsInUseTableModel.java [hunk 18]
```diff
@@ -400,4 +409,24 @@ public Component getTableCellRendererComponent(JTable table, Object value, boole
             return renderButton;
         }
     }
+
+    @Override
+    public void setValueAt(Object value, int rowIndex, int columnIndex) {
```
Comment (IllianiCBT) [inline MekHQ/src/mekhq/gui/model/PartsInUseTableModel.java:414]: Do we want to do this, or would it be better to use a JSpinner?

File: gradle/wrapper/gradle-wrapper.jar
Comment (IllianiCBT) [inline gradle/wrapper/gradle-wrapper.jar:1]: Did you mean to include this?

File: MekHQ/userdata/data/universe/ranks.xml
Comment (IllianiCBT) [inline MekHQ/userdata/data/universe/ranks.xml:1]: I could have sworn we already addressed this change? Clearly not, and it's inclusion isn't an issue.

**Review by IllianiCBT — CHANGES_REQUESTED** [review]
I'm really looking forward to having this merged, however there are a couple of points of feedback.

The biggest ones are relating to weird whitespacing that doesn't match stylistically to the rest of the mhq codebase, and abbreviated variable names. This might seem petty, but we need to remain consistent where possible to make it easy for other contributors to drop in without having to learn multiple styles.

**Comment by Lapras** [comment]
> I'm really looking forward to having this merged, however there are a couple of points of feedback.
> 
> The biggest ones are relating to weird whitespacing that doesn't match stylistically to the rest of the mhq codebase, and abbreviated variable names. This might seem petty, but we need to remain consistent where possible to make it easy for other contributors to drop in without having to learn multiple styles.

I've got no problem going through all these style changes, it's my first PR to the repo so I expected to have to readjust style a lot. I'm currently travelling, but I'll be able to readjust all of this over the weekend no problem.

File: MekHQ/src/mekhq/campaign/Campaign.java (hunk 11 above)
Comment (Lapras) [reply MekHQ/src/mekhq/campaign/Campaign.java:8492]: Just checked this, Ammunition is not a Bundle when it's made a part in use, only armor is, and this is an edge case exclusively for armor, comment was misleading, it's been revised.

File: MekHQ/src/mekhq/campaign/io/CampaignXmlParser.java (hunk 14 above)
Comment (Lapras) [reply MekHQ/src/mekhq/campaign/io/CampaignXmlParser.java]: Don't know where it came from, maybe the IDE, I did not edit this, I've reverted it in recent changes.

File: gradle/wrapper/gradle-wrapper.jar
Comment (Lapras) [reply gradle/wrapper/gradle-wrapper.jar:1]: This is not an intended include, it's something that gradle cooked up for the dev environment that might not already be in the gitignore

File: MekHQ/userdata/data/universe/ranks.xml
Comment (Lapras) [reply MekHQ/userdata/data/universe/ranks.xml:1]: This might just be because my pull/fork was from longer ago, unsure of where it's coming from

File: MekHQ/src/mekhq/gui/dialog/PartsReportDialog.java (hunk 17 above)
Comment (Lapras) [reply MekHQ/src/mekhq/gui/dialog/PartsReportDialog.java]: these sysstem.outs are for my own internal debugging and I forgot to remove them, they've all been purged from the code

File: MekHQ/src/mekhq/gui/model/PartsInUseTableModel.java (hunk 18 above)
Comment (Lapras) [reply MekHQ/src/mekhq/gui/model/PartsInUseTableModel.java:414]: Being able to type it in fits a lot better into the table layout, and the typing + the regex makes it pretty easy to edit, unless someone really wants the change, the typeable box works very well.

Comment (IllianiCBT) [reply MekHQ/src/mekhq/gui/model/PartsInUseTableModel.java:414]: Solid reasoning :)

File: MekHQ/src/mekhq/campaign/Campaign.java (hunk 6 above)
Comment (Lapras) [reply MekHQ/src/mekhq/campaign/Campaign.java]: this has been changed to partInUseRequestedStockMap, it's long winded but it's a decent descriptor of what's otherwise a difficult to describe concept. The comment above it'd declaration should help somewhat.

**Comment by Lapras** [comment]
Rather than responding to each one I'll comment here. I've implemented all the requested style changes and combed for all of the areas where they needed to be implemented. Although I might have missed a few spots. I've responded individually to other issues like some specific variable naming + system.outs, everything has been cleaned up and it should be ready now.

**Comment by HammerGS** [comment]
Looks like we have merge conflict
